vendor/blockchain_contracts/tests/rfc003_bitcoin_htlc.rs

```diff
@@ -202,14 +217,14 @@ fn redeem_htlc_with_long_secret() -> Result<(), failure::Error> {
 
     let container = docker.run(BitcoinCore::default());
     let client = tc_bitcoincore_client::new(&container);
-    client.generate(101, None).unwrap();
+    client.generate(101, None)?;
```

Comment:
We discussed that with @tcharding. I found this feature (using `?` in tests) quite cool when released but I remember that it was not as useful as it does not display the full stack when failing and hence is less useful than `unwrap()` to investigate tests.

Because I was not sure, I told Tobin to just check himself and keep the one that provide the most useful message.

@tcharding do you remember this discussion? What are the different stack traces?

Reply:
After testing, I think we should always use `unwrap()` in tests.

I have modified the following line: https://github.com/comit-network/comit-rs/blob/ae1474239084116cba8740d84c8cae431570cf40/vendor/bitcoin_witness/tests/sign_with_rate.rs#L37

Replacing `rate` with `0` (also need to make the function return a `Result<(),bitcoincore_rpc::Error>`).

This fails line 41.

If line 41 has a `unwrap()` we get the following trace:
```
thread 'sign_with_rate' panicked at 'called `Result::unwrap()` on an `Err` value: JsonRpc(Rpc(RpcError { code: -26, message: "min relay fee not met, 0 < 111 (code 66)", data: None }))', src/libcore/result.rs:997:5
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
stack backtrace:
   0: std::sys::unix::backtrace::tracing::imp::unwind_backtrace
             at src/libstd/sys/unix/backtrace/tracing/gcc_s.rs:39
   1: std::sys_common::backtrace::_print
             at src/libstd/sys_common/backtrace.rs:71
   2: std::panicking::default_hook::{{closure}}
             at src/libstd/sys_common/backtrace.rs:59
             at src/libstd/panicking.rs:197
   3: std::panicking::default_hook
             at src/libstd/panicking.rs:208
   4: <std::panicking::begin_panic::PanicPayload<A> as core::panic::BoxMeUp>::get
             at src/libstd/panicking.rs:474
   5: std::panicking::continue_panic_fmt
             at src/libstd/panicking.rs:381
   6: std::panicking::try::do_call
             at src/libstd/panicking.rs:308
   7: <T as core::any::Any>::type_id
             at src/libcore/panicking.rs:85
   8: core::result::unwrap_failed
             at /rustc/3c235d5600393dfe6c36eeed34042efad8d4f26e/src/libcore/macros.rs:18
   9: core::result::Result<T,E>::unwrap
             at /rustc/3c235d5600393dfe6c36eeed34042efad8d4f26e/src/libcore/result.rs:798
  10: sign_with_rate::sign_with_rate
             at vendor/bitcoin_witness/tests/sign_with_rate.rs:41 #<< This is the line where the unwrap fails.
  11: sign_with_rate::sign_with_rate::{{closure}}
             at vendor/bitcoin_witness/tests/sign_with_rate.rs:10 #<< This is where the test is defined
  12: core::ops::function::FnOnce::call_once
             at /rustc/3c235d5600393dfe6c36eeed34042efad8d4f26e/src/libcore/ops/function.rs:231
  13: <alloc::boxed::Box<F> as core::ops::function::FnOnce<A>>::call_once
             at /rustc/3c235d5600393dfe6c36eeed34042efad8d4f26e/src/liballoc/boxed.rs:702
  14: panic_unwind::dwarf::eh::read_encoded_pointer
             at src/libpanic_unwind/lib.rs:87
  15: test::run_test::run_test_inner::{{closure}}
             at /rustc/3c235d5600393dfe6c36eeed34042efad8d4f26e/src/libstd/panicking.rs:272
             at /rustc/3c235d5600393dfe6c36eeed34042efad8d4f26e/src/libstd/panic.rs:388
             at src/libtest/lib.rs:1468
```

If we use `?` we get the following:
```
Error: JsonRpc(Rpc(RpcError { code: -26, message: "min relay fee not met, 0 < 111 (code 66)", data: None }))


Expected :0
Actual   :1
<Click to see difference>

thread 'sign_with_rate' panicked at 'assertion failed: `(left == right)`
  left: `1`,
 right: `0`: the test returned a termination value with a non-zero status code (1) which indicates a failure', src/libtest/lib.rs:335:5
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
stack backtrace:
   0: std::sys::unix::backtrace::tracing::imp::unwind_backtrace
             at src/libstd/sys/unix/backtrace/tracing/gcc_s.rs:39
   1: std::sys_common::backtrace::_print
             at src/libstd/sys_common/backtrace.rs:71
   2: std::panicking::default_hook::{{closure}}
             at src/libstd/sys_common/backtrace.rs:59
             at src/libstd/panicking.rs:197
   3: std::panicking::default_hook
             at src/libstd/panicking.rs:208
   4: <std::panicking::begin_panic::PanicPayload<A> as core::panic::BoxMeUp>::get
             at src/libstd/panicking.rs:474
   5: std::panicking::continue_panic_fmt
             at src/libstd/panicking.rs:381
   6: std::panicking::try::do_call
             at src/libstd/panicking.rs:336
   7: test::assert_test_result
             at /Users/froyer/src/comit-rs/<::std::macros::panic macros>:8
   8: sign_with_rate::sign_with_rate::{{closure}}
             at vendor/bitcoin_witness/tests/sign_with_rate.rs:10 #<< This is where the test is defined
   9: core::ops::function::FnOnce::call_once
             at /rustc/3c235d5600393dfe6c36eeed34042efad8d4f26e/src/libcore/ops/function.rs:231
  10: <alloc::boxed::Box<F> as core::ops::function::FnOnce<A>>::call_once
             at /rustc/3c235d5600393dfe6c36eeed34042efad8d4f26e/src/liballoc/boxed.rs:702
  11: panic_unwind::dwarf::eh::read_encoded_pointer
             at src/libpanic_unwind/lib.rs:87
  12: test::run_test::run_test_inner::{{closure}}
             at /rustc/3c235d5600393dfe6c36eeed34042efad8d4f26e/src/libstd/panicking.rs:272
             at /rustc/3c235d5600393dfe6c36eeed34042efad8d4f26e/src/libstd/panic.rs:388
             at src/libtest/lib.rs:1468
```

As you can see, when using `?`, the stack only point to where the test is defined, (as one would expect), it does not point to where the unwrap fails.

Meaning that if a test fail, we would need to look at everywhere `?` is used in the test, check what `?` returns the Error type printed, etc. Which is much more work.

Hence, I'd like to propose that as a convention we only use `unwrap()` in tests. I know I am the one who introduced `?` with the migration to Rust 2018 but it was a mistake 🙇🏻‍♂️

Reply:
Fair!

I'll change it back :)
Something to announce in slack maybe? :D

Reply:
If someone else in the team agrees then yes, it'd go in slack annoucements :D

Or is 2 of us a consensus?

Reply:
Good catch.

Reply:
> If someone else in the team agrees then yes, it'd go in slack announcements.
> 
> Or is 2 of us a consensus? :D

Well, @tcharding seems to have been involved as-well, so it is at least a majority :D

Reply:
> > If someone else in the team agrees then yes, it'd go in slack announcements.
> > Or is 2 of us a consensus? :D
> 
> Well, @tcharding seems to have been involved as-well, so it is at least a majority :D

I don;t know what was @tcharding's opinion at the end.